src/main/java/ch/njol/skript/lang/function/FunctionRegistry.java

```diff
@@ -239,35 +306,77 @@ public Function<?> getFunction(@Nullable String namespace, @NotNull String name,
 		Preconditions.checkNotNull(namespace, "namespace cannot be null");
 		Preconditions.checkNotNull(provided, "provided cannot be null");
 
 		Namespace ns = namespaces.getOrDefault(namespace, new Namespace());
 		Set<FunctionIdentifier> existing = ns.identifiers.get(provided.name);
 		if (existing == null) {
 			Skript.debug("No functions named '%s' exist in the '%s' namespace".formatted(provided.name, namespace.name));
-			return null;
+			return new FunctionRetrieval(RetrievalResult.NOT_REGISTERED, null, null);
 		}
 
 		Set<FunctionIdentifier> candidates = candidates(provided, existing);
 		if (candidates.isEmpty()) {
 			Skript.debug("Failed to find a function for '%s'".formatted(provided.name));
-			return null;
+			return new FunctionRetrieval(RetrievalResult.NOT_REGISTERED, null, null);
 		} else if (candidates.size() == 1) {
 			Skript.debug("Matched function for '%s': %s".formatted(provided.name, candidates.stream().findAny().orElse(null)));
-			return ns.functions.get(candidates.stream().findAny().orElse(null));
+			return new FunctionRetrieval(RetrievalResult.EXACT,
+				ns.functions.get(candidates.stream().findAny().orElse(null)),
+				null);
 		} else {
 			String options = candidates.stream().map(Record::toString).collect(Collectors.joining(", "));
 			Skript.debug("Failed to match an exact function for '%s'".formatted(provided.name));
 			Skript.debug("Identifier: %s".formatted(provided));
 			Skript.debug("Options: %s".formatted(options));
-			return null;
+			return new FunctionRetrieval(RetrievalResult.AMBIGUOUS,
+				null,
+				candidates.stream()
+					.map(FunctionIdentifier::args)
+					.toArray(Class<?>[][]::new));
 		}
 	}
 
+	/**
+	 * The result of trying to retrieve a signature.
+	 * <p>
+	 * When getting a signature, the following situations may occur.
+	 * These are specified by {@code result}.
+	 * <ul>
+	 *     <li>
+	 *         {@code NOT_REGISTERED}. The specified signature is not registered.
+	 *         Both {@code signature} and {@code conflictingArgs} will be null.
+	 *     </li>
+	 *     <li>
+	 *         {@code AMBIGUOUS}. There are multiple signatures that
+	 *         may fit the provided name and argument types.
+	 *         {@code signature} will be null, and {@code conflictingArgs}
+	 * 		   will contain the conflicting signature parameters.
+	 *     </li>
+	 *     <li>
+	 *         {@code EXACT}. A single signature has been found which matches the name and argument types.
+	 *         {@code signature} will contain the signature, and {@code conflictingArgs} will be null.
+	 *     </li>
+	 * </ul>
+	 * </p>
+	 *
+	 * @param result          The result of the function retrieval.
+	 * @param signature       The signature that was found if {@code result} is {@code EXACT}.
+	 * @param conflictingArgs The conflicting arguments if {@code result} is {@code AMBIGUOUS}.
+	 */
+	public record SignatureRetrieval(
```

Comment:
do function retrieval and signature need to be separate records?

Reply:
i'm not sure how to make it use the same class without using an Object for the type, or is it okay in this case?

Reply:
Ah i overlooked that, seems fine then

Reply:
Well actually couldn't you use a generic?
Retrival<Function> and Retrival<Signature>